test/test_linalg.py

```diff
@@ -17,21 +17,143 @@
+        zero_strided = torch.randn(1).to(device=device, dtype=dtype).expand(50)
+        run_test_case(zero_strided, b)
+        run_test_case(a, zero_strided)
+
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    @precisionOverride({torch.bfloat16: 1e-1})
+    @dtypes(*(torch.testing.get_all_dtypes()))
+    def test_addr(self, device, dtype):
+        def run_test_case(m, a, b, beta=1, alpha=1):
+            if dtype == torch.bfloat16:
+                a_np = a.to(torch.double).cpu().numpy()
+                b_np = b.to(torch.double).cpu().numpy()
+                m_np = m.to(torch.double).cpu().numpy()
+            else:
+                a_np = a.cpu().numpy()
+                b_np = b.cpu().numpy()
+                m_np = m.cpu().numpy()
+
+            if beta == 0:
+                expected = alpha * np.outer(a_np, b_np)
+            else:
+                expected = beta * m_np + alpha * np.outer(a_np, b_np)
+
+            self.assertEqual(torch.addr(m, a, b, beta=beta, alpha=alpha), expected)
+            self.assertEqual(torch.Tensor.addr(m, a, b, beta=beta, alpha=alpha), expected)
+
+            result_dtype = torch.addr(m, a, b, beta=beta, alpha=alpha).dtype
+            out = torch.empty_like(m, dtype=result_dtype)
+            torch.addr(m, a, b, beta=beta, alpha=alpha, out=out)
+            self.assertEqual(out, expected)
+
+        a = torch.randn(50).to(device=device, dtype=dtype)
+        b = torch.randn(50).to(device=device, dtype=dtype)
+        m = torch.randn(50, 50).to(device=device, dtype=dtype)
+
+        # when beta is zero
+        run_test_case(m, a, b, beta=0., alpha=2)
+
+        # when beta is not zero
+        run_test_case(m, a, b, beta=0.5, alpha=2)
+
+        # test transpose
+        m_transpose = torch.transpose(m, 0, 1)
+        run_test_case(m_transpose, a, b, beta=0.5, alpha=2)
+
+        # test 0 strided tensor
+        zero_strided = torch.randn(1).to(device=device, dtype=dtype).expand(50)
+        run_test_case(m, zero_strided, b, beta=0.5, alpha=2)
+
+        # test scalar
+        m_scalar = torch.tensor(1, device=device, dtype=dtype)
+        run_test_case(m_scalar, a, b)
+
+    @dtypes(*itertools.product(torch.testing.get_all_dtypes(),
+                               torch.testing.get_all_dtypes()))
+    def test_outer_type_promotion(self, device, dtypes):
+        a = torch.randn(5).to(device=device, dtype=dtypes[0])
+        b = torch.randn(5).to(device=device, dtype=dtypes[1])
+        for op in (torch.outer, torch.Tensor.outer, torch.ger, torch.Tensor.ger):
+            result = op(a, b)
+            self.assertEqual(result.dtype, torch.result_type(a, b))
+
+    @dtypes(*itertools.product(torch.testing.get_all_dtypes(),
+                               torch.testing.get_all_dtypes()))
+    def test_addr_type_promotion(self, device, dtypes):
+        a = torch.randn(5).to(device=device, dtype=dtypes[0])
+        b = torch.randn(5).to(device=device, dtype=dtypes[1])
```

Comment:
This test is good. Technically it could take a triple of dtypes (a, b, c) and check the result type is:

```
torch.result_type(torch.result_type(a, b), c)
```

Reply:
According to the doc, 
https://github.com/pytorch/pytorch/blob/32c1a8c79ff8c1f69f7a798c95207963a0d47d96/torch/_torch_docs.py#L6400-L6402

Inputs have to be `tensor` instead of `dtype`. we may not able to use `torch.result_type(torch.result_type(a, b), c)` for a triple of types as the first input is dtype.

Reply:
Good point. You might have to write it as 

```
torch.promote_types(torch.result_type(a, b), c.dtype)
```